concordium-node/src/bin/cli.rs

```diff
@@ -485,33 +488,126 @@ async fn maybe_do_out_of_band_catchup(
+    }
+}
+
+// An index entry for a chunk of blocks. Its format must correspond to one
+// produced by `database-exporter`.
+#[derive(serde::Deserialize)]
+struct BlockChunkData {
+    // exported chunk of blocks' filename
+    filename:           String,
+    // genesis block index from which relative heights of blocks in the chunk are counted
+    genesis_index:      usize,
+    // relative height of the oldest block stored in the chunk
+    #[allow(dead_code)]
+    first_block_height: u64,
+    // relative height of the newest block stored in the chunk
+    last_block_height:  u64,
+}
+
+async fn import_missing_blocks(
+    consensus: &ConsensusContainer,
+    import_stopped: Arc<atomic::AtomicBool>,
+    genesis_block_hashes: &[HashBytes],
+    index_url: &url::Url,
+    data_dir_path: &Path,
+) -> anyhow::Result<()> {
+    let current_genesis_index = genesis_block_hashes.len() - 1;
+    let last_finalized_block_height = consensus.get_last_finalized_block_height();
+
+    trace!("Current genesis index: {}", current_genesis_index);
+    trace!("Local last finalized block height: {}", last_finalized_block_height);
+
+    let http_client = Client::builder().build()?;
+    let index_response = http_client.get(index_url.clone()).send().await?;
+    anyhow::ensure!(
+        index_response.status() == StatusCode::OK,
+        "Unable to download the catchup index file from {}: {} {}",
+        index_url,
+        index_response.status().as_str(),
+        index_response.status().canonical_reason().unwrap()
+    );
+
+    let mut index_reader = index_response
+        .bytes_stream()
+        .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))
+        .into_async_read();
+    let mut first_line: String = String::new();
+    index_reader.read_line(&mut first_line).await?;
+
+    let index_genesis_block_hash = first_line
+        .strip_prefix("# genesis hash ")
+        .context(
+            "The catchup index file does not begin with a line containing the genesis block hash. \
+             Please verify that you specified a correct catchup service URL. If the specified URL \
+             is correct, contact the catchup service administrator.",
+        )?
+        .trim();
+    let genesis_hash = genesis_block_hashes[0].to_string();
+    anyhow::ensure!(
+        index_genesis_block_hash == genesis_hash,
+        "The genesis block hash in the catchup index file {} does not match the genesis block \
+         hash {} in the local tree state. Please verify that you chose the catchup service for \
+         the correct chain.",
+        index_genesis_block_hash,
+        genesis_hash
+    );
+
+    let mut chunk_records = csv_async::AsyncReaderBuilder::new()
+        .comment(Some(b'#'))
+        .has_headers(false)
+        .create_deserializer(index_reader)
+        .into_deserialize();
+    while let Some(result) = chunk_records.next().await {
+        if import_stopped.load(atomic::Ordering::Acquire) {
+            break;
```

Comment:
This is not entirely correct. In case the import is stopped while consensus has control the logs will say 
```
"Out of band catchup stopped."
```
but in case it is stopped while, e.g., parsing the index file, the logs will contain
```
"Completed out of band catch-up from ..."
```

It would be consistent to instead return an Err here so that the control flows are consistent.

```rust
ensure!(!import_stopped.load(atomic::Ordering::Acquire), "Import stopped by the user.")
```